ledger-core/virtual/integration/messagediff/messagediff.go

```diff
@@ -0,0 +1,421 @@
+		// If the field is time.Time, use Equal to compare
+		if typ.String() == "time.Time" {
+			aTime := aVal.Interface().(time.Time)
+			bTime := bVal.Interface().(time.Time)
+			if !aTime.Equal(bTime) {
+				d.Modified[&localPath] = &Different{aVal.Interface().(time.Time).String(), bVal.Interface().(time.Time).String()}
+				equal = false
+			}
+		} else {
+			for i := 0; i < typ.NumField(); i++ {
+				index := []int{i}
+				field := typ.FieldByIndex(index)
+				if field.Tag.Get("testdiff") == "ignore" { // skip fields marked to be ignored
+					continue
+				}
+				if _, skip := opts.ignoreField[field.Name]; skip {
+					continue
+				}
+				localPath := append(localPath, StructField(field.Name))
+				aI := unsafeReflectValue(aVal.FieldByIndex(index))
+				bI := unsafeReflectValue(bVal.FieldByIndex(index))
+				if eq := d.diff(aI, bI, localPath, opts); !eq {
+					equal = false
+				}
+			}
+		}
+	case reflect.Ptr:
+		equal = d.diff(aVal.Elem(), bVal.Elem(), localPath, opts)
+
+	// ///////////////////////////////////////////////////////////////////
+	// Primitive kinds
+	// ///////////////////////////////////////////////////////////////////
+
+	case reflect.Float32, reflect.Float64:
+		// Round floats to FloatPrecision decimal places to compare with
+		// user-defined precision. As is commonly know, floats have "imprecision"
+		// such that 0.1 becomes 0.100000001490116119384765625. This cannot
+		// be avoided; it can only be handled. Issue 30 suggested that floats
+		// be compared using an epsilon: equal = |a-b| < epsilon.
+		// In many cases the result is the same, but I think epsilon is a little
+		// less clear for users to reason about. See issue 30 for details.
+		floatFormat := fmt.Sprintf("%%.%df", 10)
+		aval := fmt.Sprintf(floatFormat, aVal.Float())
+		bval := fmt.Sprintf(floatFormat, bVal.Float())
+		if aval != bval {
+			d.Modified[&localPath] = &Different{aVal.Float(), bVal.Float()}
+			equal = false
+		}
+	case reflect.Bool:
+		if aVal.Bool() != bVal.Bool() {
+			d.Modified[&localPath] = &Different{aVal.Bool(), bVal.Bool()}
+			equal = false
+		}
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		if aVal.Int() != bVal.Int() {
+			d.Modified[&localPath] = &Different{aVal.Int(), bVal.Int()}
+			equal = false
+		}
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		if aVal.Uint() != bVal.Uint() {
+			d.Modified[&localPath] = &Different{aVal.Uint(), bVal.Uint()}
+			equal = false
+		}
+	case reflect.String:
+		if aVal.String() != bVal.String() {
+			d.Modified[&localPath] = &Different{aVal.String(), bVal.String()}
+			equal = false
+		}
+	case reflect.Interface:
+
+		a := aVal.Interface()
+		b := bVal.Interface()
+
+		if !reflect.DeepEqual(a, b) {
+
+			fmt.Printf("\nFind interface %v %v %v", localPath, a, aType)
+			fmt.Printf("\nFind interface %v %v %v %v", localPath, b, bVal.Type(), bVal.MethodByName("Equal"))
+
+			if aType.Implements(equalType) {
```

Comment:
Type of Reference, Any, Binary is interface. aType.Implements(equalType) return false